proposals/0004-nio-redis.md

```diff
@@ -0,0 +1,217 @@
+
+`NIORedis` will provide the essential structures for interacting with RESP requests and responses, with an encoder, decoder, and representations of RESP data as `RESPValue`.
+
+In addition, SwiftNIO Channel Handlers will be provided for common tasks such as `RESPEncoder`, `RESPDecoder`, and `RedisCommandHandler`.
+
+All types defined in the package will be public with some parts being open for extension. This is to provide much of the "out of the box" experience,
+ while still leaving it entirely possible to customize the exact interaction behaviors with Redis.
+
+Server Metrics and Logging API implementations can be passed in order to receive the respective information at runtime.
+
+Full example of use:
+
+```swift
+import NIORedis
+
+let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+let driver = NIORedisDriver(ownershipModel: .external(eventLoopGroup))
+
+// connections
+
+// passing a value to 'password' will automatically authenticate with Redis before resolving the connection
+let connection = try driver.makeConnection(
+    hostname: "localhost", // this is the default
+    port: 6379, // this is the default
+    password: "MY_PASS" // default is 'nil'
+).wait()
+
+// convenience methods for commands
+
+let result = try connection.set("my_key", to: "some value")
+    .flatMap { return connection.get("my_key") }
+    .wait()
+print(result) // Optional("some value")
+
+// raw commands
+
+let keyCount = try connection.command("DEL", [RESPValue(bulk: "my_key")])
+    .flatMapThrowing { response in
+        guard case let .integer(count) = response else {
+            // throw error
+        }
+        return count
+    }
+    .wait()
+print(keyCount) // 1
+
+// cleanup
+
+connection.close()
+    .flatMapThrowing { try driver.terminate() }
+    .whenSuccess { try eventLoopGroup.syncShutdownGracefully() }
+```
+
+## Detailed Design
+
+### RESP
+
+[`RESPValue`](https://github.com/Mordil/nio-redis/blob/master/Sources/NIORedis/RESP/RESPValue.swift) will represent the different types outlined in Redis' protocol as an **enum**.
+
+```swift
+public enum RESPValue {
+    case null
+    case simpleString(String)
+    case bulkString(Data)
+    case error(RedisError)
+    case integer(Int)
+    case array([RESPValue])
+}
+```
+
+A protocol for translating Swift types to RESP is provided as [`RESPConvertible`](https://github.com/Mordil/nio-redis/blob/master/Sources/NIORedis/RESP/RESPConvertible.swift).
+ Default conformance is provided for most Swift primitives.
+
+```swift
+public protocol RESPConvertible {
+    init?(_ value: RESPValue)
+
+    func convertedToRESP() -> RESPValue
+}
+```
+
+The [`RESPEncoder`](https://github.com/Mordil/nio-redis/blob/master/Sources/NIORedis/ChannelHandlers/RESPEncoder.swift) is a class that conforms to the `MessageToByteEncoder` NIO protocol that will translate `RESPValue`s into a RESP byte stream.
+
+```swift
+public final class RESPEncoder {
+    public func encode(_ value: RESPValue) -> Data
```

Comment:
I think it could be more performant to serialize directly to `ByteBuffer`. @weissi would probably know more.

Reply:
Well, this method isn’t intended for ByteBuffer serialization. This is a general purpose method of Encoding to RESP data.

In the `MessageToByteEncoder` conformance the encode method is called and directly written to the ByteBuffer.

Again the NIO team would know more, but I’m assuming the optimizer would elide the Data directly to the ByteBuffer with as few allocations as possible